common/persistence/sql/sqlTaskManager.go

```diff
@@ -239,51 +239,98 @@ func (m *sqlTaskManager) UpdateTaskQueue(request *persistence.UpdateTaskQueueReq
 }
 
 type taskQueuePageToken struct {
-	ShardID     int
-	NamespaceID string
-	Name        string
-	TaskType    int64
+	MinRangeHash   uint32
+	MinNamespaceID primitives.UUID
+	MinName        string
+	MinTaskType    int64
 }
 
 func (m *sqlTaskManager) ListTaskQueue(request *persistence.ListTaskQueueRequest) (*persistence.ListTaskQueueResponse, error) {
-	pageToken := taskQueuePageToken{TaskType: math.MinInt16, NamespaceID: minUUID}
+	pageToken := taskQueuePageToken{MinTaskType: math.MinInt16, MinNamespaceID: minUUID}
 	if request.PageToken != nil {
 		if err := gobDeserialize(request.PageToken, &pageToken); err != nil {
 			return nil, serviceerror.NewInternal(fmt.Sprintf("error deserializing page token: %v", err))
 		}
 	}
 	var err error
 	var rows []sqlplugin.TaskQueuesRow
-	namespaceID := primitives.MustParseUUID(pageToken.NamespaceID)
-	for pageToken.ShardID < m.nShards {
-		rows, err = m.db.SelectFromTaskQueues(&sqlplugin.TaskQueuesFilter{
-			ShardID:                pageToken.ShardID,
-			NamespaceIDGreaterThan: &namespaceID,
-			NameGreaterThan:        &pageToken.Name,
-			TaskTypeGreaterThan:    &pageToken.TaskType,
-			PageSize:               &request.PageSize,
+	var shardGreaterThan uint32
+	var shardLessThan uint32
+
+	// Last page was full if these were set, see page token creation below
+	lastPageFull := pageToken.MinName != "" || pageToken.MinNamespaceID != nil
+	if lastPageFull {
+		// Since last page was full, for completeness we need to explicitly drain the last range_hash seen.
+		// A better option would be to always add one to PageSize internally to peek ahead as this results in extra queries to ensure we can move forward.
+		// todo: revisit @shawn
+		rows, err =  m.db.SelectFromTaskQueues(&sqlplugin.TaskQueuesFilter{
+			RangeHash: 				     pageToken.MinRangeHash,
+			NamespaceIDGreaterThan:      &pageToken.MinNamespaceID,
+			NameGreaterThan:             &pageToken.MinName,
+			TaskTypeGreaterThan:         &pageToken.MinTaskType,
+			PageSize:                    &request.PageSize,
 		})
 		if err != nil {
 			return nil, serviceerror.NewInternal(err.Error())
 		}
+
+		// Increment the shard as we exhausted this hash in the above query.
+		pageToken.MinRangeHash++
+	}
+
+	i := uint32(0)
+	if pageToken.MinRangeHash > 0 {
+		// Resume partition position from page token, if exists, before entering loop
+		i = getPartitionForRangeHash(pageToken.MinRangeHash, m.taskScanPartitions)
+	}
+
+	for ; i < m.taskScanPartitions; i++ {
 		if len(rows) > 0 {
 			break
 		}
-		pageToken = taskQueuePageToken{ShardID: pageToken.ShardID + 1, TaskType: math.MinInt16, NamespaceID: minUUID}
+
+		// Get start/end boundaries for partition
+		shardGreaterThan, shardLessThan = getBoundariesForPartition(i, m.taskScanPartitions)
```

Comment:
Looks like you do inclusive queries on the rangeHash, so I'm assuming edges among consecutive shards will overlap.  This could theoretically result in duplicate records but since you send the query to a different shard it does not matter as the hash range will only exist in one partition.  
But still I think it is worth calling out here in comments for future maintainer.

Reply:
Changed this to no longer overlap